### DISCUSSION TITLE
Bochs running ROM of a real BIOS

### DISCUSSION THREAD

**Discussion by sskras** [bochs-emu/Bochs#153] [Ideas]
#148 inspired a thought: could Bochs run ROM of a real BIOS?
Just like PCem or 86Box does:

[![image](https://github.com/bochs-emu/Bochs/assets/7887758/6b888196-4988-4bac-b300-9c0597964b87)](https://tasvideos.org/EmulatorResources/PCem/DOS/Configurations#Bios_2)

[![image](https://github.com/bochs-emu/Bochs/assets/7887758/44c6bf33-557c-49d4-944a-b194783c06a7)](https://tasvideos.org/EmulatorResources/PCem/DOS/Configurations#Bios_3)

[![image](https://github.com/bochs-emu/Bochs/assets/7887758/9f978269-ff32-4fe4-8a46-47dfe9ca4e4e)](https://github.com/86Box/86Box/issues/781#issuecomment-640115805)

[![image](https://github.com/bochs-emu/Bochs/assets/7887758/85712912-2464-4080-9de8-c9046f598bb7)](https://rs1.es/tutorials/2022/02/19/86box-emulator.html#configure-bios-to-detect-hard-disk)

[![image](https://github.com/bochs-emu/Bochs/assets/7887758/ae335e36-05d4-4675-af44-0951a4d3c9ea)](https://www.erikmoberg.net/article/retro-pc-emulation-with-86box#:~:text=The%20BIOS%20setup.)

Some BIOS ROMs: https://github.com/86Box/roms/tree/master/machines

**Comment by stlintel** [comment]
Yes, at some day I was able to run and boot Win95 with real asus_p6np5.bin

**Comment by stlintel** [comment]
![image](https://github.com/bochs-emu/Bochs/assets/77239907/1a7fcef1-4f96-4eb2-8329-98975320e953)

**Comment by stlintel** [comment]
But it dies a little later without being able to recognize Windows 98 boot disk

![image](https://github.com/bochs-emu/Bochs/assets/77239907/ba3f6d21-829e-46b2-9ffa-93e81df60101)

**Comment by stlintel** [comment]
Actually now booted it successfully after entering to the setup and going though HDD auto detection

![image](https://github.com/bochs-emu/Bochs/assets/77239907/8c925781-8d71-425d-9220-93c31612f571)

**Comment by fysnet** [comment]
I too got Bochs to boot Win98 with a very similar Asus BIOS.
By default, the BIOS boots `A,C` and I have to (as you did) set the `HDD Auto Detection`.
Since Bochs doesn't save these settings on exit from the BIOS, you have to reset these every time you boot.

Therefore, I should use the the `cmosimage: file=cmos.img, rtc_init=time0` option, so that the CMOS state is saved on exit.
However, if I don't have the `cmos.img` file created, Bochs panics. Maybe rather than panic, it should gracefully create the file and load the default CMOS values as if the file was never specified. This would allow you to save them on exit.
If not, an example `cmos.img` file should be provided, correct?

Does this sound like a reasonable addition (I am willing to add) or am I missing something?

**Reply by stlintel** [reply]
if you just press 'continue' after PANIC it works ?
Because Bochs knows how to act if there is no cmosimage option specified at all.
I believe if cmos.img file is not found it should act the same.
Eventually the file will be created and will be found.

**Reply by vruppert** [reply]
It's not only the CMOS RAM. A real BIOS writes configuration data into the flash ROM, so we would need to write back data to the ROM image. Using save/restore doesn't help, since it only saves the RAM contents.

**Comment by fysnet** [comment]
> if you just press 'continue' after PANIC it works ?

No, it doesn't. I had tried this too.
https://github.com/bochs-emu/Bochs/blob/9216d228be0691b518cb41f1de2ea43a87e0f5d5/bochs/iodev/cmos.cc#L202-L214
If file not found, at line 208 it Panics. If I hit continue, at line 211 it tries to get the `fstat` of the opened file. Since it cannot, it should panic again at line 213. However, in Win10, the `fstat` function crashes Bochs instead of returning error.

> It's not only the CMOS RAM. A real BIOS writes configuration data into the flash ROM

Indeed this is true, but I think we still need to do something about the CMOS binary file.

I think it would be a simple task to save/restore the BIOS/CMOS images as well as save the BIOS image on exit. A .bochsrc config flag could even be used to indicate to do this or not.

Just my thoughts on the matter.
Thanks

**Comment by vruppert** [comment]
I have now added the CMOS image option to the runtime menu. Now you can start without an image and then save the image on exit. In addition to this, we possibly need an "autosave" parameter to start with an image and not saving it on exit.

**Comment by vruppert** [comment]
Please have a look at pull request #162. It adds save/restore support for the ESCD space of the flash BIOS ROM.

**Comment by fysnet** [comment]
Thank you. I will have a look.

**Comment by vruppert** [comment]
Pull request #162 is already merged and I have now added the parameter 'flash_data' to the romimage option. Flash BIOS config data is loaded if file exists and saved on exit if modified.
